File: 2024/AoC24-14ab.rs
use std::fs;
use std::hash::{Hash, Hasher};
use std::collections::HashSet;
use std::ops::{Add, AddAssign, Mul};

const BOUNDS: Point = Point {
  x: 101,
  y: 103,
};

fn main() {
  let input = fs::read_to_string("input.txt").unwrap();

  let robots: Vec<Robot> = process(&input);

  println!("Part 1 answer: {}", part1(robots.to_owned()));
  println!("Part 2 answer: {}", part2(robots.to_owned()));
}

fn part1(mut robots: Vec<Robot>) -> usize {
  for r in &mut robots { r.step(100); }

  [
    |&r: &&Robot| r.pos.x < BOUNDS.x / 2 && r.pos.y < BOUNDS.y / 2,
    |&r: &&Robot| r.pos.x < BOUNDS.x / 2 && r.pos.y > BOUNDS.y / 2,
    |&r: &&Robot| r.pos.x > BOUNDS.x / 2 && r.pos.y < BOUNDS.y / 2,
    |&r: &&Robot| r.pos.x > BOUNDS.x / 2 && r.pos.y > BOUNDS.y / 2,
  ].into_iter().map(|f| robots.iter().filter(f).count()).product()
}

fn part2(mut robots: Vec<Robot>) -> u32 {
  let mut n = 0;
  while HashSet::<Robot>::from_iter(robots.to_owned()).len() != robots.len() {
    n += 1;
    for r in &mut robots { r.step(1); }
  }

  n
}

fn process(input: &str) -> Vec<Robot> {
  input.lines().map(|line| {
    let mut t = line
      .split(|c: char| !"-0123456789".contains(c))
      .filter_map(|s|
        if !s.is_empty() { Some(s.parse::<i32>().unwrap()) }
        else { None }
      );
    assert_eq!(t.to_owned().count(), 4);

    Robot {
      pos: Point {
        x: t.next().unwrap(),
        y: t.next().unwrap(),
      },
      vel: Point {
        x: t.next().unwrap(),
        y: t.next().unwrap(),
      },
    }
  }).collect()
}

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
struct Point {
  x: i32,
  y: i32,
}

impl Point {
  fn rem_euclid(self, rhs: Self) -> Self {
    Self {
      x: self.x.rem_euclid(rhs.x),
      y: self.y.rem_euclid(rhs.y),
    }
  }
}

impl Add for Point {
  type Output = Self;
  fn add(self, other: Self) -> Self {
    Self {
      x: self.x + other.x,
      y: self.y + other.y,
    }
  }
}

impl AddAssign for Point {
  fn add_assign(&mut self, other: Self) {
    *self = *self + other;
  }
}

impl Mul<i32> for Point {
  type Output = Self;
  fn mul(self, n: i32) -> Self {
    Self {
      x: self.x * n,
      y: self.y * n,
    }
  }
}

#[derive(Clone, Copy)]
struct Robot {
  pos: Point,
  vel: Point,
}

impl Robot {
  fn step(&mut self, n: i32) {
    self.pos += self.vel * n;
    self.pos = self.pos.rem_euclid(BOUNDS);
  }
}

impl PartialEq for Robot {
  fn eq(&self, other: &Self) -> bool {
    self.pos == other.pos
  }
}

impl Eq for Robot {}

impl Hash for Robot {
  fn hash<H: Hasher>(&self, state: &mut H) {
    self.pos.hash(state);
  }
}

File: 2024/AoC24-14ab.dart
import 'dart:io';

const int WIDTH = 101;
const int HEIGHT = 103;

void main() {
  final String input = File('input.txt').readAsStringSync();
  List<Robot> robots = input.lines.map(process).toList();

  print('Part 1 answer: ${part1(robots.clone())}');
  print('Part 2 answer: ${part2(robots.clone())}');
}

int part1(List<Robot> robots) {
  for (Robot r in robots) r.step(100);

  final int midX = WIDTH ~/ 2;
  final int midY = HEIGHT ~/ 2;
  
  return <bool Function(Robot)>[
    (r) => r.pos.x < midX && r.pos.y < midY,
    (r) => r.pos.x < midX && r.pos.y > midY,
    (r) => r.pos.x > midX && r.pos.y < midY,
    (r) => r.pos.x > midX && r.pos.y > midY,
  ].fold<int>(1, (acc, v) => acc * robots.where(v).length);
}

int part2(List<Robot> robots) {
  int i = 0;
  while (true) {
    i += 1;

    for (Robot r in robots) r.step();
    if (Set.of(robots).length == robots.length) return i;
  }
}

Robot process(String line) {
  List<int> temp = line
    .split(RegExp(r'[^\d|\-]+'))
    .where((s) => s.isNotEmpty)
    .map(int.parse)
    .toList();
  if (temp.length != 4) exit(1);

  return Robot(Point(temp[0], temp[1]), Point(temp[2], temp[3]));
}

class Point {
  int x, y;

  Point(this.x, this.y);
  Point.clone(Point p) : this(p.x, p.y);

  Point operator +(Point other) =>
    Point(this.x + other.x, this.y + other.y);

  Point operator *(int n) =>
    Point(this.x * n, this.y * n);

  Point operator %(Point other) =>
    Point(this.x % other.x, this.y % other.y);

  @override
  bool operator ==(Object other) =>
    other is Point && this.x == other.x && this.y == other.y;

  @override
  int get hashCode => Object.hash(x, y);
}

class Robot {
  Point pos;
  Point vel;

  Robot(this.pos, this.vel);
  Robot.clone(Robot r) : this(Point.clone(r.pos), Point.clone(r.vel));

  void step([int time = 1]) {
    this.pos += (this.vel * time);
    this.pos %= Point(WIDTH, HEIGHT);
  }

  @override
  bool operator ==(Object other) =>
    other is Robot && this.pos == other.pos;

  @override
  int get hashCode => pos.hashCode;
}

extension on String {
  Iterable<String> get lines => this.split('\n');
}

extension <E> on List<E> {
  List<E> clone() => List.generate(this.length, (i) => this[i]);
}